### PR TITLE
Check dict item accesses where it isn't already checked

### PR DESCRIPTION
**Pull request by virtuald**
## Description

* Initial stab at https://github.com/pybind/pybind11/issues/2862

## Suggested changelog entry:

N/A

### CODE REVIEW

File: include/pybind11/pytypes.h [hunk 1]
```diff
@@ -478,6 +478,42 @@ inline handle get_function(handle value) {
...
+    if (rv == NULL && PyErr_Occurred()) {
+        throw error_already_set();
+    }
+    return rv;
+#else
+    return PyDict_GetItemString(v, key);
+#endif
+}
+
+inline PyObject * dict_getitem(PyObject *v, PyObject *key)
```
Comment (bstaletic) [inline include/pybind11/pytypes.h:504]: Why isn't this a member of `py::dict`?

Comment (virtuald) [reply include/pybind11/pytypes.h:504]: Because the code that uses it uses the raw CPython API, presumably for performance reasons?

**Comment by virtuald** [comment]
Also, if someone could retrigger the checks for this, it's likely that the tests will pass.

File: include/pybind11/pytypes.h (hunk 1 above)
Comment (bstaletic) [reply include/pybind11/pytypes.h:504]: Fair enough.

File: include/pybind11/pytypes.h [hunk 2]
```diff
@@ -478,6 +478,42 @@ inline handle get_function(handle value) {
...
+inline PyObject * dict_getitemstring(PyObject *v, const char *key)
+{
+#if PY_MAJOR_VERSION >= 3
+    PyObject *kv, *rv;
+    kv = PyUnicode_FromString(key);
+    if (kv == NULL) {
+        throw error_already_set();
+    }
+
+    rv = PyDict_GetItemWithError(v, kv);
```
Comment (bstaletic) [inline include/pybind11/pytypes.h:493]: At this point, can't you just call `dict_getitemstring`?

Comment (virtuald) [reply include/pybind11/pytypes.h:493]: Nope, still has potential to raise an exception. See https://bugs.python.org/issue35459

As copied from cpython @ https://github.com/python/cpython/blob/d5fc99873769f0d0d5c5d5d99059177a75a4e46e/Objects/dictobject.c#L1545

Comment (bstaletic) [reply include/pybind11/pytypes.h:493]: My bad, I was thinking about calling your `dict_getitem`. From this line on, `dict_getitemstring` and `dict_getitem` are the same.

Comment (virtuald) [reply include/pybind11/pytypes.h:493]: Not quite the same, Py_DECREF needs to be called on the key before the error is thrown.

File: include/pybind11/pytypes.h [hunk 3]
```diff
@@ -478,6 +478,42 @@ inline handle get_function(handle value) {
     return value;
 }
 
+// Reimplementation of python's dict helper functions
+
+// copied from cpython _PyDict_GetItemStringWithError
+inline PyObject * dict_getitemstring(PyObject *v, const char *key)
```
Comment (EricCousineau-TRI) [inline include/pybind11/pytypes.h:484]: nit Can you summarize rationale for this? (and also for traceability?)
e.g. "We should use `*WithError` flavors to ensure we don't accidentally shadow other signals like KeyboardInterrupt (#2862)" or sthg like that? (either ref'ing your issue, or https://bugs.python.org/issue35459)?

**Review by EricCousineau-TRI — APPROVED** [review]
Looks great from my side! Just a minor nit on rationale comment. (Yeah, discoverable via `git blame` + clicks, but mayhaps a terse statement could help!)

**Comment by rwgk** [comment]
Hi Dustin, sorry for the long delay! This PR escaped me before. It looks very important.
Could you please rebase this PR to trigger the CI? That will run new clang-tidy checks we added last week.

File: include/pybind11/pytypes.h (hunk 3 above)
Comment (virtuald) [reply include/pybind11/pytypes.h:492]: Done.

**Comment by rwgk** [comment]
Nice, thanks! Good to see that the new clang-tidy is happy.

**Comment by rwgk** [comment]
I forgot to mention, I already integrated this PR for Google-internal global testing. I'll hold off merging until I get the results, later tonight if things go well.

**Comment by rwgk** [comment]
The Google-internal testing was successful. Merging!
